src/Bls12381G1KeyPair.ts

```diff
@@ -0,0 +1,506 @@
+        throw new Error("No private key to sign with.");
+      }
+    };
+  }
+  return {
+    async sign({ data }): Promise<Uint8Array> {
+      //TODO assert data runtime Uint8Array | Uint8Array[]
+      if (data instanceof Uint8Array) {
+        return blsSign({
+          messages: [data],
+          keyPair: {
+            secretKey: new Uint8Array(key.privateKeyBuffer as Uint8Array),
+            publicKey: new Uint8Array(key.publicKeyBuffer)
+          }
+        });
+      }
+      return blsSign({
+        messages: data,
+        keyPair: {
+          secretKey: new Uint8Array(key.privateKeyBuffer as Uint8Array),
+          publicKey: new Uint8Array(key.publicKeyBuffer)
+        }
+      });
+    }
+  };
+};
+
+/**
+ * NOTE: This function isn't properly producing signatures at this time. An update
+ * needs to be made to the bbs-signatures library before this API will work properly
+ * for Bls12381G1KeyPairs.
+ *
+ * @ignore
+ * Returns an object with an async verify function for verifying BBS+ signatures.
+ * The verify function is bound to the KeyPair
+ * and then returned by the KeyPair's verifier method.
+ * @param key - A Bls12381G1KeyPair.
+ *
+ * @returns An async verifier specific
+ * to the key passed in.
+ */
+const verifierFactory = (key: Bls12381G1KeyPair): KeyPairVerifier => {
+  if (!key.publicKeyBuffer) {
+    return {
+      async verify(): Promise<boolean> {
+        throw new Error("No public key to verify with.");
+      }
+    };
+  }
+
+  return {
+    async verify({ data, signature }): Promise<boolean> {
+      //TODO assert data instance of Uint8Array | Uint8Array[]
+      if (data instanceof Uint8Array) {
+        return blsVerify({
+          messages: [data],
+          publicKey: new Uint8Array(key.publicKeyBuffer),
+          signature
+        }).verified;
+      }
+      return blsVerify({
+        messages: data,
+        publicKey: new Uint8Array(key.publicKeyBuffer),
+        signature
+      }).verified;
+    }
+  };
+};
+
+/**
+ * A BLS 12-381 G1 based key pair
+ */
+export class Bls12381G1KeyPair {
+  /**
+   * The key id
+   */
+  readonly id?: string;
+  /**
+   * The key controller
+   */
+  readonly controller?: string;
+  /**
+   * Buffer containing the raw bytes of the private key
+   */
+  readonly privateKeyBuffer?: Uint8Array;
+  /**
+   * Buffer containing the raw bytes of the public key
+   */
+  readonly publicKeyBuffer: Uint8Array;
+  /**
+   * Type identifier for the key pair
+   */
+  readonly type: string = "Bls12381G1Key2020";
+
+  /**
+   * Default constructor.
+   */
+  constructor(options: KeyPairOptions) {
+    /**
+     * The provided publicKey needs to be 384 bits / 5.85 = 65.6
+     * which means the base58 encoded publicKey can be either 65 or 66 chars
+     * 5.85 = log base 2 (58) which is equivalent to the number of bits
+     * encoded per character of a base58 encoded string.
+     *
+     */
+    if (
+      options.publicKeyBase58.length !== 65 &&
```

Comment:
Just a general question, are there any other examples of public keys that are variable length e.g 65 or 66

Reply:
I'd expect anything that's bae58 encoded would encounter this variability in key length. Base64/base64url are as well if padding is dropped I suppose, but the variability is less likely because it's bit length per character is a whole integer.

The general solution for if something is variable is key length in bits modulo log base 2 (encoding character set). If there's a remainder then it will be variable, but if it's 0 it will always be one length.

So for G1 keys with base64 encoding we've got 384 bits modulo log 2(64) = 0 

For G2 with base64 encoding it's 768 bits mod log2(64) = 0. 

However by switching to base58 we change the dividend from 6 to 5.85 making it impossible for it to be not variable because it's a decimal dividend and the divisor is always an integer.